File: Solutions/Java/src/test/java/StudentTest.java
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import static org.junit.jupiter.api.Assertions.*;

class StudentTest {

    @Nested
    class StudentClassTests {
        Student mikey;

        @BeforeEach
        void setUp() {
            mikey = new Student("Mikey", List.of(85, 90, 78));
        }

        @Test
        void shouldCalculateAverageGradeCorrectly() {
            assertEquals(84.33, mikey.getAverageGrade(), 0.01);
        }

        @Test
        void shouldReturnCorrectLetterGrade() {
            assertEquals("B", mikey.getLetterGrade());
        }
    }

    @Nested
    class StudentTrackerClassTests {
        StudentTracker tracker;

        @BeforeEach
        void setUp() {
            tracker = new StudentTracker();
        }

        @Test
        void shouldAddStudentToTracker() {
            tracker.addStudent("Andy", List.of(92, 88, 95));
            assertEquals(1, tracker.getStudentCount());
        }

        @Test
        void shouldGetStudentsWithSpecificLetterGrade() {
            Student mikey = tracker.addStudent("Mikey", List.of(95, 92, 98));
            Student chunk = tracker.addStudent("Chunk", List.of(75, 70, 68));
            Student mouth = tracker.addStudent("Mouth", List.of(65, 61, 58));
            Student brand = tracker.addStudent("Brand", List.of(85, 82, 88));
            Student stef = tracker.addStudent("Stef", List.of(91, 89, 93));
            Student troy = tracker.addStudent("Troy", List.of(55, 62, 59));

            assertEquals(2, tracker.getStudentsWithLetterGrade("A").size());
            assertTrue(tracker.getStudentsWithLetterGrade("A").contains(mikey));
            assertTrue(tracker.getStudentsWithLetterGrade("A").contains(stef));

            assertEquals(1, tracker.getStudentsWithLetterGrade("B").size());
            assertTrue(tracker.getStudentsWithLetterGrade("B").contains(brand));

            assertEquals(1, tracker.getStudentsWithLetterGrade("C").size());
            assertTrue(tracker.getStudentsWithLetterGrade("C").contains(chunk));

            assertEquals(1, tracker.getStudentsWithLetterGrade("D").size());
            assertTrue(tracker.getStudentsWithLetterGrade("D").contains(mouth));

            assertEquals(1, tracker.getStudentsWithLetterGrade("F").size());
            assertTrue(tracker.getStudentsWithLetterGrade("F").contains(troy));
        }
    }
}

File: Solutions/Kotlin/src/test/kotlin/StudentTest.kt
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class StudentTest {

    @Nested
    inner class `Student Class` {
        @Test
        fun `should calculate the average grade correctly`() {
            val mikey = Student("Mikey", listOf(85, 90, 78))
            assertEquals(84.33, mikey.averageGrade, 0.01)
        }

        @Test
        fun `should return the correct letter grade`() {
            val mikey = Student("Mikey", listOf(95, 92, 98))
            val brand = Student("Brand", listOf(80, 88, 76))
            val chunk = Student("Chunk", listOf(75, 70, 68))
            val mouth = Student("Mouth", listOf(65, 61, 58))
            val troy = Student("Troy", listOf(55, 62, 59))

            assertEquals("A", mikey.letterGrade)
            assertEquals("B", brand.letterGrade)
            assertEquals("C", chunk.letterGrade)
            assertEquals("D", mouth.letterGrade)
            assertEquals("F", troy.letterGrade)
        }
    }

    @Nested
    inner class `StudentTracker Class` {
        @Test
        fun `should add a student to the tracker`() {
            val tracker = StudentTracker()
            tracker.addStudent("Andy", listOf(92, 88, 95))
            assertEquals(1, tracker.studentCount)
        }

        @Test
        fun `should get students with a specific letter grade`() {
            val tracker = StudentTracker()
            val mikey = tracker.addStudent("Mikey", listOf(95, 92, 98))
            val chunk = tracker.addStudent("Chunk", listOf(75, 70, 68))
            val mouth = tracker.addStudent("Mouth", listOf(65, 61, 58))
            val brand = tracker.addStudent("Brand", listOf(85, 82, 88))
            val stef = tracker.addStudent("Stef", listOf(91, 89, 93))
            val troy = tracker.addStudent("Troy", listOf(55, 62, 59))

            val gradeAStudents = tracker.getStudentsWithLetterGrade("A")
            val gradeBStudents = tracker.getStudentsWithLetterGrade("B")
            val gradeCStudents = tracker.getStudentsWithLetterGrade("C")
            val gradeDStudents = tracker.getStudentsWithLetterGrade("D")
            val gradeFStudents = tracker.getStudentsWithLetterGrade("F")

            assertEquals(2, gradeAStudents.size)
            assertEquals(true, gradeAStudents.contains(mikey))
            assertEquals(true, gradeAStudents.contains(stef))

            assertEquals(1, gradeBStudents.size)
            assertEquals(true, gradeBStudents.contains(brand))

            assertEquals(1, gradeCStudents.size)
            assertEquals(true, gradeCStudents.contains(chunk))

            assertEquals(1, gradeDStudents.size)
            assertEquals(true, gradeDStudents.contains(mouth))

            assertEquals(1, gradeFStudents.size)
            assertEquals(true, gradeFStudents.contains(troy))
        }
    }
}
